### ISSUE TITLE
upgrade to EF core 3

### ISSUE THREAD

**Issue by NickLydon** [PawelGerr/Thinktecture.EntityFrameworkCore#3]
Hi! 
Thanks for your work on this. I tried to leave a message on the azure devops project page, but have no idea if that's possible.
We're using the project for the rownumber support and at the moment it's preventing us upgrading to EF core 3. Is there anything we can do to help get a release out?

**Comment by PawelGerr** [comment]
Hi, at the moment there is no way to create issue in azure devops that's why I keep this github-repo :)

Usually, I release something after using the lib in own projects for some time but at the moment all our customers who need this lib are still using EF 2.x. 

Are you using `2.0.0-beta005` right now?

**Comment by mantasaudickas** [comment]
Hi,

Our case works only with 2.0.0-beta001.

2.0.0-beta005 is not working and throws an exception:
```
System.InvalidOperationException: The LINQ expression 'DbSet<RoleEntity>
    .Where(r => (Nullable<DateTime>)r.CreateDate >= __searchParameters_NewerThan_0)
    .Select(r => new {
        RoleId = r.RoleId,
        RowNumber = __Functions_1
            .RowNumber(__Functions_1
                .OrderBy(r.Name)
                .ThenBy(r.RoleId))
     })
    .AsSubQuery()
    .Where(e => e.RoleId == __id_2)' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to either AsEnumerable(), AsAsyncEnumerable(), ToList(), or ToListAsync(). See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.<VisitMethodCall>g__CheckTranslated|8_0(ShapedQueryExpression translated, <>c__DisplayClass8_0& )
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Thinktecture.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Thinktecture.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass9_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at System.Linq.Queryable.SingleOrDefault[TSource](IQueryable`1 source, Expression`1 predicate)
   at LinqToDB.AsyncExtensions.<>c__DisplayClass19_0`1.<SingleOrDefaultAsync>b__0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location where exception was thrown ---
   at Authorization.RoleRepository.RowNumberAsync(IQueryable`1 query, Guid id, ISortParameters`1 searchParameters) in d:\dev\authorizationservice\Authorization.Repositories\RoleRepository.cs:line 208
```

Code we are using:
```
var rowNumbers = query.Select(x => new
{
    x.RoleId,
    RowNumber = EF.Functions.RowNumber(EF.Functions.OrderBy(x.Name).ThenBy(x.RoleId))
})

var row = await  AsyncExtensions.SingleOrDefaultAsync(
    RelationalQueryableExtensions.AsSubQuery(rowNumbers),
    x => x.RoleId == id
).ConfigureAwait(false);
return row?.RowNumber;
```

**Comment by PawelGerr** [comment]
Did you call `AddRowNumberSupport` on the `DbContextOptionsBuilder`?
Are you using SQL Server or SQLite?

```csharp
.AddDbContext<DemoDbContext>(builder => builder
                              .UseSqlServer(ConnectionString)
                              .AddRowNumberSupport());

```

I've tried with a similar query but don't get any exceptions. Can you provide me with a small repro of this issue?

**Comment by mantasaudickas** [comment]
Yes, AddRowNumberSupport is there:
```
var migrationsAssembly = typeof(AuthorizationDbContext).GetTypeInfo().Assembly.GetName().Name;

services
    .AddDbContext<AuthorizationDbContext>(options =>
        options
            .UseSqlServer(connectionString, sql =>
            {
                sql.MigrationsAssembly(migrationsAssembly);
            }).AddRowNumberSupport()
    );
```

We will try to make a repro..  however - project is not very small, so might take some time.

**Comment by NickLydon** [comment]
Hi @PawelGerr , sorry for the delay. I've added [a repro](https://github.com/NickLydon/ThinktectureRowNumberIssue). The issue is on the master branch with version beta005, but you can switch to the [version-beta001 branch](https://github.com/NickLydon/ThinktectureRowNumberIssue/tree/version-beta001) to see how it worked successfully with one of the pre-releases.

**Comment by PawelGerr** [comment]
@NickLydon Thank you for your support. I've released a new version `2.0.0-beta006`, pls test the changes with your real project. 
The root cause had something to do with the configuration, i.e. `AddRowNumberSupport`. To prevent such configuration-issues I moved `AddRowNumberSupport` _closer to the database_. That way I know exactly what DB it is.

![image](https://user-images.githubusercontent.com/7135790/83550216-c911df80-a506-11ea-9808-4ceec265e948.png)

The [docs](https://dev.azure.com/pawelgerr/Thinktecture.EntityFrameworkCore/_wiki/wikis/Thinktecture.EntityFrameworkCore.wiki/14/RowNumber-Support?anchor=1.-activate-the-rownumber-support) are updated accordingly.
